### PR TITLE
2269 publish events from hadoop exploiter

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Fixes part of #2269.

Update Hadoop exploiter to publish propagation and exploitation events.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [x] ~Was the CHANGELOG.md updated to reflect the changes?~
* [x] ~~Was the documentation framework updated to reflect the changes?~~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] ~~Added relevant unit tests?~~
* [x] ~~Have you successfully tested your changes locally? Elaborate:~~
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] ~~If applicable, add screenshots or log transcripts of the feature working~~

image

### CODE REVIEW

File: monkey/infection_monkey/exploit/hadoop.py [hunk 1]
```diff
@@ -51,7 +57,7 @@ def _exploit_host(self):
                 target=self.host.ip_addr,
                 exploitation_success=False,
                 error_message=self.exploit_result.error_message,
-                tags=(HADOOP_EXPLOITER_TAG,),
+                tags=PROPAGATION_TAGS,
```
Comment (ilija-lazoroski) [inline monkey/infection_monkey/exploit/hadoop.py:60]: Why propagation tags when we publish an exploitation event?

Comment (cakekoa) [reply monkey/infection_monkey/exploit/hadoop.py]: Good catch -- fixed!

File: monkey/infection_monkey/exploit/hadoop.py [hunk 2]
```diff
@@ -90,7 +126,15 @@ def exploit(self, url, command):
...
         )
-        return resp.status_code == 202
+
+        success = resp.status_code == 202
+        if success:
+            self._publish_exploitation_event(
+                target=self.host.ip_addr,
+                exploitation_success=True,
+                tags=EXPLOIT_TAGS,
+            )
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/hadoop.py:136]: Timestamps need to be recorded as close to the actual event as possible.

Comment (mssalvatore) [reply monkey/infection_monkey/exploit/hadoop.py:136]: What about on failure?

```suggestion
        self._publish_exploitation_event(
            target=self.host.ip_addr,
            exploitation_success=success,
            tags=EXPLOIT_TAGS,
        )
```

File: monkey/infection_monkey/exploit/hadoop.py [hunk 3]
```diff
@@ -40,12 +52,16 @@ def _exploit_host(self):
         urls = self.build_potential_urls(self.host.ip_addr, self.HADOOP_PORTS)
         self.add_vulnerable_urls(urls, True)
         if not self.vulnerable_urls:
+            self.exploit.error_message = f"No vulnerable urls has been found for {self.host}"
+            self._publish_exploitation_event(
+                target=self.host.ip_addr,
+                exploitation_success=False,
+                error_message=self.exploit_result.error_message,
+                tags=EXPLOIT_TAGS,
+            )
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/hadoop.py:61]: I think just do this in `exploit()`. See also comments on timestamps.

File: monkey/infection_monkey/exploit/hadoop.py [hunk 4]
```diff
@@ -58,13 +74,33 @@ def _exploit_host(self):
...
+                )
+                self._publish_propagation_event(
+                    self.host.ip_addr, False, PROPAGATION_TAGS, error_message
+                )
+        except requests.RequestException as err:
+            error_message = str(err)
+            self._publish_exploitation_event(self.host.ip_addr, False, EXPLOIT_TAGS, error_message)
+            self._publish_propagation_event(
+                self.host.ip_addr, False, PROPAGATION_TAGS, error_message
+            )
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/hadoop.py:96]: Timestamps

**Review by mssalvatore — CHANGES_REQUESTED** [review]
Left some questions. A big question, though, is this:

Do we need to check for vulnerable URLs? Why not just exploit stuff until we succeed? It seems unnecessary. Maybe @VakarisZ has some insight on the history.

**Comment by VakarisZ** [comment]
> Left some questions. A big question, though, is this:
> 
> Do we need to check for vulnerable URLs? Why not just exploit stuff until we succeed? It seems unnecessary, since exploit and check do basically the same thing with a different command. Maybe @VakarisZ has some insight on the history.

I don't think there's any reason besides consistency. If we want to change it we will need to reeavaluate exploitation indicators. I don't remember if a response to exploitation request needs to be 200 or it can be 500. I also don't remember if all or only exploited responses include `application-id`

**Review by shreyamalviya — COMMENTED** [review]
Same comments as https://github.com/guardicore/monkey/pull/2401

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/exploit/hadoop.py [hunk 5]
```diff
@@ -32,29 +43,32 @@ class HadoopExploiter(WebRCE):
...
         http_path, http_thread = HTTPTools.create_locked_transfer(
             self.host, str(monkey_path_on_victim), self.agent_binary_repository
         )
 
+        command = self._build_command(monkey_path_on_victim, http_path)
         try:
-            command = self._build_command(monkey_path_on_victim, http_path)
 
-            if self.exploit(self.vulnerable_urls[0], command):
+            if self.exploit(potential_urls[0], command):
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/hadoop.py:71]: We need to try all potential URLs.

**Review by mssalvatore — CHANGES_REQUESTED** [review]

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**